_posts/2022-03-01-auto-backport-in-opensearch.markdown

```diff
@@ -0,0 +1,46 @@
+
+In this blog post I would like to talk about the challenges that came along while integrating auto-backports and the approaches taken to resolve them incrementally.
+
+1. The backport workflow did not have the permissions to create pull requests from a backport branch using the GITHUB_TOKEN. This happened since the GITHUB_TOKEN has read permissions for all scopes unless specified otherwise in the workflow. This was resolved by adding a [`permissions`](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token) key in the workflow file in order to provide the required scope (in this case related to `pull-requests`). Additionally, these updated permissions do not work when used on `pull_request` target events since such events can only get write permissions if the PR is within the source repo; cross-repo PRs do not get write permissions. For this purpose, [`pull_request_target`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target) event can be used.
+
+2. The original [backport](https://github.com/tibdex/backport) Github Action created backport branches of the format backport-${pullRequestNumber}-to-${base} which conflicted with other branch protection rules. To maintain branch permissions for such custom branches, we ended up adding a [custom branch name](https://github.com/VachaShah/backport#features) feature on the Github Action which allows passing a branch name prefix as an input, so if the branch name prefix passed is `backport/backport-${pullRequestNumber}`, the workflow will create a branch named `backport/backport-${pullRequestNumber}-to-${base}` for auto-backports.
+ 
+3. For pull requests generated automatically for backports, it is also necessary to have all the required CI workflows running on them similar to a normal pull request. This helps validate the backport pull requests against release branches preventing an incompatible merge. However, according to Github's [documentation](https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#triggering-a-workflow-from-a-workflow), when an action in a workflow is performed using the `GITHUB_TOKEN`, it cannot trigger a new workflow run. This means that when the backport Github Action runs on the original pull request and creates a backport pull request as a result of the workflow, CI workflows cannot be triggered on such pull requests. To solve this problem, we created a simple Github App with required permissions so it can work with the backport workflow to create backport pull requests instead of using the `GITHUB_TOKEN`.
```

Comment:
- 'it is also necessary to have all the required CI workflows running on them similar to a normal pull request' -> 'it is necessary to have all the required continuous integration (CI) workflows running just as a normal pull request'
- 'according to Github's' -> 'according to GitHub's'
- 'This means that when the backport Github Action' -> 'This means that when the backport GitHub Action'
- 'creates a backport pull request as a result of the workflow, CI workflows cannot be triggered on such pull requests.' -> 'creates a backport pull request from the workflow, another workflows cannot be triggered.'
- 'To solve this problem, we created a simple Github App with required permissions so it can work with the backport workflow to create backport pull requests instead of using the `GITHUB_TOKEN`.' -> 'Creating a simple GitHub App with the required permissions solved this problem by allowing it to make the backport pull requests instead of using the `GITHUB_TOKEN`'